### PR TITLE
executor: add diagnosis rule to check some metrics exceed thresholds

### PR DESCRIPTION
**Pull request by crazycs520**
Signed-off-by: crazycs <crazycs520@gmail.com>

### What problem does this PR solve? 
This PR adds a new diagnosis rule -- `threshold-check`, which is used to check some metrics threshold value, like tikv thread cpu usage. We will detect the following metrics tables in the current implementation:

* `tikv_thread_cpu`

Consider the PR length, this PR only add some part checks of `threshold-check` rule, I will add more check in the next PR.

### Check List 

Tests 

 - Unit test
 - Manual test 

Release note

 - Add diagnosis rule `threshold-check` which is used to check some metrics threshold value.

### CODE REVIEW

File: executor/diagnostics.go [hunk 1]
```diff
@@ -79,13 +79,17 @@ type (
 	// criticalErrorInspection is used to check are there some critical errors
 	// occurred in the past
 	criticalErrorInspection struct{ inspectionName }
+
+	// thresholdCheckInspection is used to check some threshold value, like cpu usage, leader count change.
```
Comment (zimulala) [inline executor/diagnostics.go]: s/cpu/CPU

File: executor/diagnostics.go [hunk 2]
```diff
@@ -363,3 +367,64 @@ func (criticalErrorInspection) inspect(ctx context.Context, sctx sessionctx.Cont
...
+		{item: "coprocessor_high_cpu", component: "cop_high%", configKey: "readpool.coprocessor.high-concurrency", threshold: 0.9},
+		{item: "coprocessor_low_cpu", component: "cop_low%", configKey: "readpool.coprocessor.low-concurrency", threshold: 0.9},
+		{item: "grpc_cpu", component: "grpc%", configKey: "server.grpc-concurrency", threshold: 0.9},
+		{item: "raftstore_cpu", component: "raftstore_%", configKey: "raftstore.store-pool-size", threshold: 0.8},
+		{item: "apply_cpu", component: "apply_%", configKey: "raftstore.apply-pool-size", threshold: 0.8},
+		{item: "storage_readpool_normal_cpu", component: "store_read_norm%", configKey: "readpool.storage.normal-concurrency", threshold: 0.9},
+		{item: "storage_readpool_high_cpu", component: "store_read_high%", configKey: "readpool.storage.high-concurrency", threshold: 0.9},
+		{item: "storage_readpool_low_cpu", component: "store_read_low%", configKey: "readpool.storage.low-concurrency", threshold: 0.9},
+		{item: "scheduler_worker_cpu", component: "sched_%", configKey: "storage.scheduler-worker-pool-size", threshold: 0.85},
+		{item: "split_check_cpu", component: "split_check", configKey: "", threshold: 0.9},
```
Comment (Deardrops) [inline executor/diagnostics.go]:
```suggestion
		{item: "split_check_cpu", component: "split_check", threshold: 0.9},
```

File: executor/diagnostics.go [hunk 3]
```diff
@@ -363,3 +367,64 @@ func (criticalErrorInspection) inspect(ctx context.Context, sctx sessionctx.Cont
...
+				"where t1.cpu > %[2]f;", rule.component, rule.threshold)
+		}
+		rows, _, err := sctx.(sqlexec.RestrictedSQLExecutor).ExecRestrictedSQLWithContext(ctx, sql)
+		if err != nil {
+			sctx.GetSessionVars().StmtCtx.AppendWarning(fmt.Errorf("execute '%s' failed: %v", sql, err))
+			continue
+		}
+		for _, row := range rows {
+			actual := fmt.Sprintf("%.2f", row.GetFloat64(1))
+			expect := ""
```
Comment (Deardrops) [inline executor/diagnostics.go]:
```suggestion
			expected := ""
```

File: executor/diagnostics_test.go [hunk 4]
```diff
@@ -166,6 +171,140 @@ func (s *diagnosticsSuite) TestInspectionResult(c *C) {
...
+	// Mock for metric table data.
+	fpName2 := "github.com/pingcap/tidb/executor/mockMetricsTableData"
+	c.Assert(failpoint.Enable(fpName2, "return"), IsNil)
+	defer func() { c.Assert(failpoint.Disable(fpName2), IsNil) }()
+
+	datetime := func(s string) types.Time {
+		t, err := types.ParseTime(tk.Se.GetSessionVars().StmtCtx, s, mysql.TypeDatetime, types.MaxFsp)
+		c.Assert(err, IsNil)
+		return t
+	}
```
Comment (Deardrops) [inline executor/diagnostics_test.go]: Extract this function, here is a [same function definition](https://github.com/pingcap/tidb/blob/5d4eb4c806d07382043d795268497071d0f6ca5a/executor/diagnostics_test.go#L176-L180)

File: executor/diagnostics.go [hunk 5]
```diff
@@ -363,3 +367,64 @@ func (criticalErrorInspection) inspect(ctx context.Context, sctx sessionctx.Cont
...
 }
+
+func (thresholdCheckInspection) inspect(ctx context.Context, sctx sessionctx.Context, filter inspectionFilter) []inspectionResult {
+	var rules = []struct {
+		item      string
+		component string
+		configKey string
+		threshold float64
+	}{
+		{item: "coprocessor_normal_cpu", component: "cop_normal%", configKey: "readpool.coprocessor.normal-concurrency", threshold: 0.9},
```
Comment (lonng) [inline executor/diagnostics.go]: The line excced 80 characters, it's better to use multiple lines form :
```go
{
  item: "",
  component: "",
}
```

File: executor/diagnostics.go [hunk 6]
```diff
@@ -363,3 +367,64 @@ func (criticalErrorInspection) inspect(ctx context.Context, sctx sessionctx.Cont
...
+		{item: "scheduler_worker_cpu", component: "sched_%", configKey: "storage.scheduler-worker-pool-size", threshold: 0.85},
+		{item: "split_check_cpu", component: "split_check", configKey: "", threshold: 0.9},
+	}
+	var results []inspectionResult
+	for _, rule := range rules {
+		var sql string
+		if len(rule.configKey) > 0 {
+			sql = fmt.Sprintf("select t1.instance, t1.cpu, t2.threshold, t2.value from "+
+				"(select instance, sum(value) as cpu from metric_schema.tikv_thread_cpu where name like '%[1]s' and time=now() group by instance) as t1,"+
+				"(select value * %[2]f as threshold, value from information_schema.cluster_config where type='tikv' and `key` = '%[3]s' limit 1) as t2 "+
```
Comment (lonng) [inline executor/diagnostics.go]:
```suggestion
				"(select value * %[2]f as threshold, value from inspection_schema.cluster_config where type='tikv' and `key` = '%[3]s' limit 1) as t2 "+
```

File: executor/diagnostics.go [hunk 7]
```diff
@@ -363,3 +367,64 @@ func (criticalErrorInspection) inspect(ctx context.Context, sctx sessionctx.Cont
...
+		}
+		for _, row := range rows {
+			actual := fmt.Sprintf("%.2f", row.GetFloat64(1))
+			expect := ""
+			if len(rule.configKey) > 0 {
+				expect = fmt.Sprintf("< %.2f, config: %v=%v", row.GetFloat64(2), rule.configKey, row.GetString(3))
+			} else {
+				expect = fmt.Sprintf("< %.2f", row.GetFloat64(2))
+			}
+			detail := fmt.Sprintf("select instance, sum(value) as cpu from tikv_thread_cpu where name like '%[1]s' and time=now() group by instance", rule.component)
```
Comment (lonng) [inline executor/diagnostics.go]: Can this `detail` apply to all results? And we should use `db.table` form in `detail` field to make the user more easily to execute this SQL.

File: executor/diagnostics_test.go [hunk 8]
```diff
@@ -166,6 +171,140 @@ func (s *diagnosticsSuite) TestInspectionResult(c *C) {
...
+			"readpool.coprocessor.low-concurrency":    "4",
+			"readpool.storage.high-concurrency":       "4",
+			"readpool.storage.low-concurrency":        "4",
+			"readpool.storage.normal-concurrency":     "4",
+			"server.grpc-concurrency":                 "8",
+			"storage.scheduler-worker-pool-size":      "6",
+		}
+		return configuration, nil
+	}
+	router.Handle("/config", fn.Wrap(mockTiKVConfig))
```
Comment (lonng) [inline executor/diagnostics_test.go]: We should use `inspection_schema.cluster_config` in inspect mode (and it's easier to mock data in `inspection_schema.cluster_config`).

File: executor/diagnostics_test.go [hunk 9]
```diff
@@ -166,17 +169,136 @@ func (s *diagnosticsSuite) TestInspectionResult(c *C) {
...
+		"tikv,127.0.0.1,readpool.storage.high-concurrency,4",
+		"tikv,127.0.0.1,readpool.storage.low-concurrency,4",
+		"tikv,127.0.0.1,readpool.storage.normal-concurrency,4",
+		"tikv,127.0.0.1,server.grpc-concurrency,8",
+		"tikv,127.0.0.1,storage.scheduler-worker-pool-size,6",
+	}
+
+	fpName := "github.com/pingcap/tidb/infoschema/mockInspectionSchemaClusterConfigData"
+	fpExpr := strings.Join(configurations, ";")
+	c.Assert(failpoint.Enable(fpName, fmt.Sprintf(`return("%s")`, fpExpr)), IsNil)
```
Comment (lonng) [inline executor/diagnostics_test.go]: It's better to reuse the exists failpoint (`mockMergeMockInspectionTables`) to mock data.

File: infoschema/inspection_schema.go [hunk 10]
```diff
@@ -59,6 +61,23 @@ func (it *inspectionSchemaTable) IterRecords(ctx sessionctx.Context, startKey kv
 		return table.ErrUnsupportedOp
 	}
 
+	failpoint.Inject("mockInspectionSchemaClusterConfigData", func(val failpoint.Value) {
```
Comment (lonng) [inline infoschema/inspection_schema.go]: This failpoint is unused, which should be removed.

**Pull request merged**